go-controller/pkg/cni/helper_linux.go

```diff
@@ -60,31 +60,21 @@ func setupNetwork(link netlink.Link, ifInfo *PodInterfaceInfo) error {
 	if err := netlink.LinkSetHardwareAddr(link, ifInfo.MAC); err != nil {
 		return fmt.Errorf("failed to add mac address %s to %s: %v", ifInfo.MAC, link.Attrs().Name, err)
 	}
-	addr := &netlink.Addr{IPNet: ifInfo.IP}
-	if err := netlink.AddrAdd(link, addr); err != nil {
-		return fmt.Errorf("failed to add IP addr %s to %s: %v", ifInfo.IP, link.Attrs().Name, err)
+	for _, ip := range ifInfo.IPs {
+		addr := &netlink.Addr{IPNet: ip}
+		if err := netlink.AddrAdd(link, addr); err != nil {
+			return fmt.Errorf("failed to add IP addr %s to %s: %v", ip, link.Attrs().Name, err)
+		}
+	}
+	for _, gw := range ifInfo.Gateways {
+		if err := ip.AddRoute(nil, gw, link); err != nil {
+			return fmt.Errorf("failed to add gateway route: %v", err)
+		}
 	}
-
-	var foundDefault bool
 	for _, route := range ifInfo.Routes {
 		if err := ip.AddRoute(route.Dest, route.NextHop, link); err != nil {
 			return fmt.Errorf("failed to add pod route %v via %v: %v", route.Dest, route.NextHop, err)
 		}
-
-		if ones, _ := route.Dest.Mask.Size(); ones == 0 {
-			foundDefault = true
```

Comment:
@danwinship is this removed for some reason? IIRC it's there to support secondary OVN networks that might claim the default route, isntead of forcing it through the main interface.

Reply:
In the normal case, `Routes` is empty and `Gateways` contains the implied default route.

If there is a secondary network attachment with a default route, then `getRoutesGatewayIP` in `pkg/ovn/pods.go` will set `Routes` to provide explicit routes to the cluster network, and leave `Gateways` unset. It doesn't add the external default route to `Routes` though because it assumes the other CNI plugin will set the default route itself.

So there is no situation where `ifInfo.Routes` contains a default route (and `MarshalPodAnnotation` and `UnmarshalPodAnnotation` both enforce that now).